### PR TITLE
Support multi-stage Dockerfiles

### LINKED ISSUE #15905: Provide multi-stage Dockerfiles for native and jvm builds
## Description

It seems the preferred way to build Docker images in Quarkus is using the CLI, however when using for example DigitalOcean Apps, which essentially is providing auto-build & deploy based on a Dockerfile specified in the project root, you are stuck with the Quarkus way.

There is a multi-stage Dockefile for native builds in the documentation https://quarkus.io/guides/building-native-image#using-a-multi-stage-docker-build - not able to use that on DigitalOcean as it timeouts after 30 min, so I´d like to do a simple JVM build instead.

The included Dockerfile.jvm does not build anything and instead relies on the CLI build triggered through Gradle / Maven and cannot be used as it is.

## Implementation ideas

Provide a **Dockerfile.multistage.native** and **Dockerfile.multistage.jvm** inside the Quarkus Quickstarter directory so people can quickly copy the file into the root and deploy their stuff without much researching or creating custom build scripts.
image

### PR DESCRIPTION
**Pull request by gastaldi**
This introduces three new `Dockerfile` files (and their respective `.dockerignore`s) to the generated projects to support [multi-staged builds](https://docs.docker.com/build/building/multi-stage/):

- `src/main/docker/Dockerfile.multistage.jvm`
- `src/main/docker/Dockerfile.multistage.native`
- `src/main/docker/Dockerfile.multistage.native-micro`

This is required in some CI environments

- Fixes #15905

### CODE REVIEW

**Comment by maxandersen** [comment]
On first look -1 this won't work for projects unless they strictly confirm and you add zero extra files. 

Can't we do this better?

**Comment by gastaldi** [comment]
We could generate a `.dockerignore.multistage` and document that file to be used when multistaged builds are necessary to avoid slowing down the build on other Dockerfiles

**Comment by gastaldi** [comment]
@maxandersen I found out that you can create a `Dockerfile.multistage.jvm.dockerignore` next to the Dockerfile (same name but with a `.dockerignore` extension) and it will use that

**Comment by maxandersen** [comment]
Ah perfect. So that could be done in devfile rather than inside this repo.

**Comment by gastaldi** [comment]
> Ah perfect. So that could be done in devfile rather than inside this repo.

I believe this is still useful for CI environments that do not use devfiles (eg. Digital Ocean, Circle CI, etc)

**Comment by maxandersen** [comment]
Sure but how will they know how to use those files ?

File: independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/dockerfiles/codestart.yml [hunk 1]
```diff
@@ -8,3 +8,5 @@ language:
           from: registry.access.redhat.com/ubi8/ubi-minimal:8.6
         native-micro:
           from: quay.io/quarkus/quarkus-micro-image:2.0
+        builder:
+          from: quay.io/quarkus/ubi-quarkus-mandrel-builder-image:22.3-java17
```
Comment (maxandersen) [inline independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/dockerfiles/codestart.yml:12]: Why this change?

File: independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/dockerfiles/base/src/main/docker/Dockerfile.tpl.qute.multistage.native-micro [hunk 2]
```diff
@@ -0,0 +1,39 @@
+####
```
Comment (maxandersen) [inline independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/dockerfiles/base/src/main/docker/Dockerfile.tpl.qute.multistage.native-micro:1]: Adding yet another docker file?

**Review by maxandersen — CHANGES_REQUESTED** [review]
These won't be usable by devfile as is will they?

File: independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/dockerfiles/base/src/main/docker/Dockerfile.tpl.qute.multistage.native-micro (hunk 2 above)
Comment (gastaldi) [reply independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/dockerfiles/base/src/main/docker/Dockerfile.tpl.qute.multistage.native-micro:1]: > Sure but how will they know how to use those files ?

By specifying the Dockerfile path (eg. https://docs.digitalocean.com/products/app-platform/reference/dockerfile/#defining-the-dockerfile)

File: independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/dockerfiles/codestart.yml (hunk 1 above)
Comment (gastaldi) [reply independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/dockerfiles/codestart.yml:12]: That's the image used as the builder

File: independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/dockerfiles/base/src/main/docker/Dockerfile.tpl.qute.multistage.native-micro (hunk 2 above)
Comment (gastaldi) [reply independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/dockerfiles/base/src/main/docker/Dockerfile.tpl.qute.multistage.native-micro:1]: > Adding yet another docker file?

I followed the pattern we already have with the native Dockerfiles but I am fine with just one native Multi staged Dockerfile

File: independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/dockerfiles/base/src/main/docker/Dockerfile.tpl.qute.multistage.native [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+###
+
+## Stage 1 : build with Maven builder image
+FROM {dockerfile.builder.from} AS build
+ARG MAVEN_BUILD_EXTRA_ARGS=
+COPY --chown=quarkus:quarkus src mvnw pom.xml /code/
+COPY --chown=quarkus:quarkus .mvn/ /code/.mvn/
+USER quarkus
+WORKDIR /code
+RUN ./mvnw -B -ntp org.apache.maven.plugins:maven-dependency-plugin:3.5.0:go-offline
```
Comment (stuartwdouglas) [inline independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/dockerfiles/base/src/main/docker/Dockerfile.tpl.qute.multistage.native:21]: Does this go-offline code actually do anything in this case? To make this pattern work I thought you need to copy the pom.xml, then run go-offline, then copy the source. This way if there is no pom.xml change docker can re-use the layer. If you have copied everything then every change will re-download. 

Also you would likely want quarkus:go-offline to actually get all the deployment side dependencies.


In this case though you are copying everything, so any code change will likely result in a full download again.

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]

**Review by cescoffier — COMMENTED** [review]
The change looks good, but we must be aware that we are adding yet another set of docker files, which quickly leads to out-of-date files (including massive CVE in the base layers). @maxandersen mentioned several times some work to generate these files instead of creating them when the application is created.

**Comment by maxandersen** [comment]
> @maxandersen I found out that you can create a `Dockerfile.multistage.jvm.dockerignore` next to the Dockerfile (same name but with a `.dockerignore` extension) and it will use that

so i tried this but I'm not seeing it working. how did you run this build?

**Comment by gastaldi** [comment]
> > @maxandersen I found out that you can create a `Dockerfile.multistage.jvm.dockerignore` next to the Dockerfile (same name but with a `.dockerignore` extension) and it will use that
> 
> so i tried this but I'm not seeing it working. how did you run this build?

@maxandersen it's in https://docs.docker.com/engine/reference/commandline/build/#use-a-dockerignore-file, but you need to have a BuildKit backend enabled. Try enabling it via `export DOCKER_BUILDKIT=1` 

This is my Docker version for reference purposes: 
```
Client:
 Cloud integration: v1.0.31
 Version:           20.10.23
 API version:       1.41
 Go version:        go1.18.10
 Git commit:        7155243
 Built:             Thu Jan 19 17:35:19 2023
 OS/Arch:           darwin/arm64
 Context:           desktop-linux
 Experimental:      true

Server: Docker Desktop 4.17.0 (99724)
 Engine:
  Version:          20.10.23
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.10
  Git commit:       6051f14
  Built:            Thu Jan 19 17:31:28 2023
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.18
  GitCommit:        2456e983eb9e37e47538f59ea18f2043c9a73640
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

File: independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/dockerfiles/base/src/main/docker/Dockerfile.tpl.qute.multistage.native (hunk 3 above)
Comment (gastaldi) [reply independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/dockerfiles/base/src/main/docker/Dockerfile.tpl.qute.multistage.native:21]: Ah, good catch. `quarkus:go-offline` is better too, I'll change to that if we all agree this PR is a good idea

**Comment by gastaldi** [comment]
Closing as we'd like to have it automatically generated
